core/include/usModuleContext.h

```diff
@@ -1,852 +1,870 @@
+   * specified in its ServiceConstants::SERVICE_ID() property); that is, the
+   * service that was registered first is returned.
+   *
+   * @param clazz The class name with which the service was registered.
+   * @return A <code>ServiceReference</code> object, or an invalid <code>ServiceReference</code> if
+   *         no services are registered which implement the named class.
+   * @throws std::logic_error If this ModuleContext is no longer valid.
+   * @throws ServiceException If no service was registered under the given class name.
+   *
+   * @see #GetServiceReferences(const std::string&, const std::string&)
+   */
+  ServiceReferenceU GetServiceReference(const std::string& clazz);
+
+  /**
+   * Returns a <code>ServiceReference</code> object for a service that
+   * implements and was registered under the specified template class argument.
+   *
+   * <p>
+   * This method is identical to GetServiceReference(const std::string&) except that
+   * the class name for the service object is automatically deduced from the template argument.
+   *
+   * @tparam S The type under which the requested service must have been registered.
+   * @return A <code>ServiceReference</code> object, or an invalid <code>ServiceReference</code> if
+   *         no services are registered which implement the type <code>S</code>.
+   * @throws std::logic_error If this ModuleContext is no longer valid.
+   * @throws ServiceException It no service was registered under the given class name.
+   * @see #GetServiceReference(const std::string&)
+   * @see #GetServiceReferences(const std::string&)
+   */
+  template<class S>
+  ServiceReference<S> GetServiceReference()
+  {
+    std::string clazz = us_service_interface_iid<S>();
+    if (clazz.empty()) throw ServiceException("The service interface class has no US_DECLARE_SERVICE_INTERFACE macro");
+    return ServiceReference<S>(GetServiceReference(clazz));
+  }
+
+  /**
+   * Returns the service object referenced by the specified
+   * <code>ServiceReferenceBase</code> object.
+   * <p>
+   * A module's use of a service is tracked by the module's use count of that
+   * service. Each time a service's service object is returned by
+   * {@link #GetService(const ServiceReference<S>&)} the context module's use count for
+   * that service is incremented by one. Each time the service is released by
+   * {@link #UngetService(const ServiceReferenceBase&)} the context module's use count
+   * for that service is decremented by one.
+   * <p>
+   * When a module's use count for a service drops to zero, the module should
+   * no longer use that service.
+   *
+   * <p>
+   * This method will always return <code>0</code> when the service
+   * associated with this <code>reference</code> has been unregistered.
+   *
+   * <p>
+   * The following steps are taken to get the service object:
+   * <ol>
+   * <li>If the service has been unregistered, <code>0</code> is returned.
+   * <li>The co
```

Comment:
It allows static bundles to be installed explicitly into the Framework. This assumes we are adhering to OSGi's spec for the ```InstallBundle```` API.